pkg/gatewayserver/io/ws/lbslns/upstream.go

```diff
@@ -400,10 +405,56 @@ func (updf *UplinkDataFrame) FromUplinkMessage(up *ttnpb.UplinkMessage, bandID s
 	return nil
 }
 
-// ToTxAcknowledgment extracts fields from the basic station TxConfirmation "dntxed" message and converts them into a TxAcknowledgment for the network server.
-func ToTxAcknowledgment(correlationIDs []string) ttnpb.TxAcknowledgment {
-	return ttnpb.TxAcknowledgment{
-		CorrelationIDs: correlationIDs,
-		Result:         ttnpb.TxAcknowledgment_SUCCESS,
+// ToTxAck implements Format.
+func (lbsLNS) ToTxAck(ctx context.Context, raw []byte, tokens io.DownlinkTokens, receivedAt time.Time) (*ttnpb.TxAcknowledgment, ws.ParsedTime, error) {
+	var (
+		txConf TxConfirmation
+		txAck  ttnpb.TxAcknowledgment
+	)
+	if err := json.Unmarshal(raw, &txConf); err != nil {
+		return nil, ws.ParsedTime{}, err
+	}
+	if cids, _, ok := tokens.Get(uint16(txConf.Diid), receivedAt); ok {
+		txAck.CorrelationIDs = cids
+		txAck.Result = ttnpb.TxAcknowledgment_SUCCESS
+	} else {
+		logger := log.FromContext(ctx)
+		logger.WithField("diid", txConf.Diid).Debug("Tx acknowledgement either does not correspond to a downlink message or arrived too late")
 	}
+	return &txAck, ws.ParsedTime{
+		XTime:   txConf.XTime,
+		RefTime: txConf.RefTime,
+	}, nil
+}
+
+// ToUplink implements Format.
+func (lbsLNS) ToUplink(ctx context.Context, raw []byte, ids ttnpb.GatewayIdentifiers, bandID string, receivedAt time.Time, msgType string) (*ttnpb.UplinkMessage, ws.ParsedTime, error) {
+	var (
+		up         *ttnpb.UplinkMessage
+		parsedTime ws.ParsedTime
+		err        error
+	)
+	switch msgType {
+	case ws.TypeUpstreamJoinRequest:
```

Comment:
This implies that websocket protocols have different messages for join request and uplink data frames, while that is also a pretty clear LBS thing. I don't think the websocket protocol handler should care about what sort of upstream message it is.